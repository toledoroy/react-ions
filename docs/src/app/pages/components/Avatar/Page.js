import React from 'react'
import PropsList from 'private/modules/PropsList'
import docs from '!!docgen!react-ions/lib/components/Avatar/Avatar'
import CodeExample from 'private/modules/CodeExample'
import styles from 'private/css/content'
import ExampleAvatar from './ExampleAvatar'
import exampleAvatarCode from '!raw!./ExampleAvatar'
import FallbackAvatar from './FallbackAvatar'
import fallbackAvatarCode from '!raw!./FallbackAvatar'
import ExampleAvatarNoFade from './ExampleAvatarNoFade'
import exampleAvatarNoFadeCode from '!raw!./ExampleAvatarNoFade'

const description = {
  avatar: 'This is the `avatar component`.',
  fallbackAvatar: 'This is the `avatar component` with fallback letters and color',
  avatarNoFade: 'This is the `avatar component` with no fade in effect.'
}

const AvatarPage = () => (
  <div>
    <div className={styles.content}>
      <div className={styles.block}>
        <CodeExample
          title='Example Avatar'
          description={description.avatar}
          markup={exampleAvatarCode}>
          <ExampleAvatar />
        </CodeExample>
        <CodeExample
          title='Fallback Avatar'
          description={description.fallbackAvatar}
          markup={fallbackAvatarCode}>
          <FallbackAvatar />
        </CodeExample>
        <CodeExample
          title='Avatar with no Fade'
          description={description.avatarNoFade}
          markup={exampleAvatarNoFadeCode}>
          <ExampleAvatarNoFade />
        </CodeExample>
      </div>
      <div className={styles.block}>
        <h3>Props</h3>
        <PropsList list={docs[0].props} />
      </div>
    </div>
  </div>
)

export default AvatarPage
