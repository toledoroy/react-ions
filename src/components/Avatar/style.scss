@import '../../styles/common/colors';

.avatar-wrapper {
  flex-grow: 1;
  flex-shrink: 0;
  opacity: 0;
  overflow: hidden;
  position: relative;
  transition: opacity .25s ease-in;

  &.loaded {
    opacity: 1;
  }

  > div {
    align-items: center;
    display: flex;
    justify-content: center;
    user-select: none;

    span {
      color: $white;
      position: relative;
      text-transform: uppercase;
      top: 2%;
    }
  }
}
