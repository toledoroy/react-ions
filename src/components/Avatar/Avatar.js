import React from 'react'
import PropTypes from 'prop-types'
import style from './style.scss'
import classNames from 'classnames/bind'

class Avatar extends React.Component {
  constructor(props) {
    super(props)
  }

  state = {
    // If image src was passed in, it is not yet loaded
    // Else, if letters were passed in, set loaded to true
    loaded: !this.props.src
  }

  static propTypes = {
    /**
     * Optional source of the image to load.
     */
    src: PropTypes.string,
    /**
     * Optional letters to display in lieu of an image.
     */
    letters: PropTypes.string,
    /**
     * Optional background for the letters.
     */
    letterBackgroundColor: PropTypes.string,
    /**
     * Optional alt text for the image
     */
    alt: PropTypes.string,
    /**
     * Optional size to constrain the image (only supports square images)
     */
    size: PropTypes.string,
    /**
     * Optional CSS class to pass to the badge.
     */
    optClass: PropTypes.string,
    /**
     * Option to turn the opacity fade off (defaults to true)
     */
    fadeIn: PropTypes.bool
  }

  static defaultProps = {
    fadeIn: true
  }

  /* If fadeIn is set to false,
   * or if it's a letter-based avatar,
   * set loaded to true, so it won't fade in */
  state = {
    loaded: !!(!this.props.fadeIn || !this.props.src),
    square: true
  }

  /* If the source changes, and fadeIn
   * is set, fade in the new avatar */
  componentWillReceiveProps = nextProps => {
    if (nextProps.src !== this.props.src && this.props.fadeIn) {
      this.setState({ loaded: false })
    }
  }

  shouldComponentUpdate = (nextProps, nextState) => {
    return (
      nextProps.letters !== this.props.letters ||
      nextProps.src !== this.props.src ||
      nextProps.size !== this.props.size ||
      nextProps.fadeIn !== this.props.fadeIn ||
      nextState.loaded !== this.state.loaded
    )
  }

  getWrapperStyle = () => {
    return {
      backgroundColor: this.props.letterBackgroundColor || this.getBackgroundColor(),
      width: this.props.size + 'px',
      height: this.props.size + 'px'
    }
  }

  getBackgroundColor = () => {
    // If no letters passed in, but a src exists
    if (!this.props.letters && this.props.src) return 'transparent'

    // If no letters passed in, return a default color
    if (!this.props.letters) return '#F93943'

    switch (this.props.letters.charAt(0).toLowerCase()) {
      case 'a':
      case 'k':
      case 'u':
        return '#F93943'
      case 'b':
      case 'l':
      case 'v':
        return '#796DE8'
      case 'c':
      case 'm':
      case 'w':
        return '#6E3FAF'
      case 'd':
      case 'n':
      case 'x':
        return '#28D397'
      case 'e':
      case 'o':
      case 'y':
        return '#ED7C5A'
      case 'f':
      case 'p':
      case 'z':
        return '#F93983'
      case 'g':
      case 'q':
        return '#F9B339'
      case 'h':
      case 'r':
        return '#6BE2F9'
      case 'i':
      case 's':
        return '#AAE667'
      case 'j':
      case 't':
        return '#ED7BE9'
      default:
        return '#F93943'
    }
  }

  getTextStyle = () => {
    return {
      fontSize: (+this.props.size) * 0.6 + 'px'
    }
  }

  getNotSquareStyle = () => {
    return {
      backgroundImage: 'url(' + this.props.src + ')',
      backgroundSize: 'cover',
      height: this.props.size + 'px'
    }
  }

  handleLoad = ({target: img}) => {
    this.setState({ loaded: true })
    if (img.naturalHeight === img.naturalWidth) {
      this.setState({ square: true})
    } else {
      this.setState({ square: false })
    }
  }

  loadAvatar = () => {
    if (this.props.src) {
      if (this.state.square) {
        return <img src={this.props.src} onLoad={this.handleLoad} alt={this.props.alt} height={this.props.size} />
      }
      if (!this.state.square) {
        return (
          <div style={this.getNotSquareStyle()} onLoad={this.handleLoad} >
            <img src={this.props.src} alt={this.props.alt} height='0' />
          </div>
        )
      }
    } else if (this.props.letters) {
      return (
        <div style={this.getWrapperStyle()}>
          <span style={this.getTextStyle()}>{this.props.letters.length <= 2 ? this.props.letters : this.props.letters.substr(0, 2)}</span>
        </div>
      )
    }
  }

  render() {
    const cx = classNames.bind(style)
    const avatarClasses = cx(style['avatar-wrapper'], (this.state.loaded ? 'loaded' : null), this.props.optClass, this.props.className)

    return (
      <div className={avatarClasses} style={this.props.size ? this.getWrapperStyle() : null}>
        {this.loadAvatar()}
      </div>
    )
  }
}

export default Avatar
