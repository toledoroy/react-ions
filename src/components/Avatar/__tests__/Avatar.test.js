import React from 'react'
import Avatar from '../Avatar'
import renderer from 'react-test-renderer'

const data = {
  src: 'https://ambassador-api.s3.amazonaws.com/uploads/avatars/1088_2016_05_04_12_16_49.jpg',
  alt: 'Cat',
  size: '100'
}

it('should shallow render an image', () => {
  const tree = renderer
    .create(<Avatar
      src={data.src}
      alt={data.alt}
      size={data.size}
    />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('should shallow render a fallback avatar', () => {
  const tree = renderer
    .create(<Avatar letters='cf' />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('should take an optClass', () => {
  const tree = renderer
    .create(<Avatar src={data.src} optClass='test' />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})