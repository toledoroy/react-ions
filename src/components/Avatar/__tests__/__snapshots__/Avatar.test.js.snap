// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should shallow render a fallback avatar 1`] = `
<div
  className="loaded"
  style={null}
>
  <div
    style={
      Object {
        "backgroundColor": "#6E3FAF",
        "height": "undefinedpx",
        "width": "undefinedpx",
      }
    }
  >
    <span
      style={
        Object {
          "fontSize": "NaNpx",
        }
      }
    >
      cf
    </span>
  </div>
</div>
`;

exports[`should shallow render an image 1`] = `
<div
  className=""
  style={
    Object {
      "backgroundColor": "transparent",
      "height": "100px",
      "width": "100px",
    }
  }
>
  <img
    alt="Cat"
    height="100"
    onLoad={[Function]}
    src="https://ambassador-api.s3.amazonaws.com/uploads/avatars/1088_2016_05_04_12_16_49.jpg"
  />
</div>
`;

exports[`should take an optClass 1`] = `
<div
  className="test"
  style={null}
>
  <img
    onLoad={[Function]}
    src="https://ambassador-api.s3.amazonaws.com/uploads/avatars/1088_2016_05_04_12_16_49.jpg"
  />
</div>
`;
